### DISCUSSION TITLE
Reading AXIS_SERVO_DRIVE tag

### DISCUSSION THREAD

**Discussion by augustofellipe** [ottowayi/pycomm3#197] [Q&A] [answered]
I'm trying to read a motion tag for a servo drive, it's a predefined data type=AXIS_CIP_DRIVE.(*edit axis_servo_drive)
I can read any other type of tag, except this one. Has anyone faced this problem before?
<img width="608" alt="Screenshot from 2021-11-30 15-04-56" src="https://user-images.githubusercontent.com/83625286/144103791-615bedff-b219-43b7-a4f7-b3f850aec8a1.png">

**Comment by ottowayi** [comment, marked as answer]
What version are you using? and what is the firmware version of the PLC?  The latest pycomm3 version contains some fixes for predefined types.  Can you enable [logging](https://docs.pycomm3.dev/en/latest/getting_started.html#logging) and share it?  Also can you send the result of `plc.data_types['AXIS_CIP_DRIVE']`?

**Reply by ottowayi** [reply]
Also, it might not be possible to read the entire struct at once.  I get a permission denied error when I try on mine.  But your example shows it got the data, but failed to parse it, so it might be a different issue.  There are some pre-defined tags that cannot be read whole, I know some of the alarm ones are like that.  Have you tried reading individual attributes?

**Reply by augustofellipe** [reply]
Thank you in advance for replying!

1. Using Python 3.8, pycomm3 1.2.5. , PLC L62 20.013
2. Reading the individual attribute: receive a BOOL tag instead of a REAL.
<img width="608" alt="read attribute" src="https://user-images.githubusercontent.com/83625286/144124078-52427d90-55a8-4108-8513-35b5d349c1f1.png">

3. plc.data_types['AXIS_SERVO_DRIVE'] don't return anything, maybe the problem is the lack of "underline"?
<img width="620" alt="debug data_type" src="https://user-images.githubusercontent.com/83625286/144123439-18f8f1a8-9be0-48ac-8423-578c9b5a28f5.png">

4. Tomorrow I can log in and send the results.


[read.txt](https://github.com/ottowayi/pycomm3/files/7628570/read.txt)
[get_tag_info.txt](https://github.com/ottowayi/pycomm3/files/7628571/get_tag_info.txt)

**Reply by ottowayi** [reply]
I think I found the issue, but the fix might not be as easy.  So far every type I've encountered the members info and member names have been of the same length, but for some reason this type says it has more members than there are names for them.  The specs do say that predefined types don't follow the same rules, so they should not be read/written as whole structs, but I've been trying to reverse engineer what I can for them.   If you read a single attribute the response has the correct data type code in the response, but the decoding function ignores that and uses what it parsed for the structure layout instead.  I didn't notice it was doing that, so if I can't get the structure layout figured out I will at least fix the decoding so that single attributes at least work. 
![image](https://user-images.githubusercontent.com/5274950/144131669-4404a3d8-5a0c-49d1-bc07-ee070725f883.png)

**Reply by ottowayi** [reply]
I just pushed a change to the develop branch that should fix it.  It works at least with my PLC and all the unit tests still pass, can you try it out as well?  If this fix does it I'll get the code cleaned up and put out a new release.

**Reply by augustofellipe** [reply]
It worked! Using develop branch like you said
<img width="616" alt="Screenshot from 2021-12-01 08-47-51" src="https://user-images.githubusercontent.com/83625286/144230735-11651b06-33ec-450c-98e1-0d53fa4dbfaa.png">

Collecting entire predefined data type:
<img width="800" alt="Screenshot from 2021-12-01 08-54-38" src="https://user-images.githubusercontent.com/83625286/144231201-28a2715d-07ce-4973-92c6-516807fda569.png">

Collecting specific attribute:
<img width="800" alt="Screenshot from 2021-12-01 08-55-12" src="https://user-images.githubusercontent.com/83625286/144231502-28e95fce-7896-4fa9-b02d-37b24a679c74.png">

Thank you for your reply! If you need help to test any feature on-field, count on me, I work on the plant floor using the entire Rockwell PLC family.

Augusto Dutra

**Reply by ottowayi** [reply]
fixed in v1.2.6
